// @azure/arm-appservice/esm/operations/certificateRegistrationProvider.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/certificateRegistrationProviderMappers";
import * as Parameters from "../models/parameters";
/** Class representing a CertificateRegistrationProvider. */
var CertificateRegistrationProvider = /** @class */ (function () {
    /**
     * Create a CertificateRegistrationProvider.
     * @param {WebSiteManagementClientContext} client Reference to the service client.
     */
    function CertificateRegistrationProvider(client) {
        this.client = client;
    }
    CertificateRegistrationProvider.prototype.listOperations = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, listOperationsOperationSpec, callback);
    };
    CertificateRegistrationProvider.prototype.listOperationsNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listOperationsNextOperationSpec, callback);
    };
    return CertificateRegistrationProvider;
}());
export { CertificateRegistrationProvider };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listOperationsOperationSpec = {
    httpMethod: "GET",
    path: "providers/Microsoft.CertificateRegistration/operations",
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.CsmOperationCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var listOperationsNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.CsmOperationCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
