// @azure/arm-appservice/esm/operations/provider.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/providerMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Provider. */
var Provider = /** @class */ (function () {
    /**
     * Create a Provider.
     * @param {WebSiteManagementClientContext} client Reference to the service client.
     */
    function Provider(client) {
        this.client = client;
    }
    Provider.prototype.getAvailableStacks = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, getAvailableStacksOperationSpec, callback);
    };
    Provider.prototype.listOperations = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, listOperationsOperationSpec, callback);
    };
    Provider.prototype.getAvailableStacksOnPrem = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, getAvailableStacksOnPremOperationSpec, callback);
    };
    Provider.prototype.getAvailableStacksNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, getAvailableStacksNextOperationSpec, callback);
    };
    Provider.prototype.listOperationsNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listOperationsNextOperationSpec, callback);
    };
    Provider.prototype.getAvailableStacksOnPremNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, getAvailableStacksOnPremNextOperationSpec, callback);
    };
    return Provider;
}());
export { Provider };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getAvailableStacksOperationSpec = {
    httpMethod: "GET",
    path: "providers/Microsoft.Web/availableStacks",
    queryParameters: [
        Parameters.osTypeSelected,
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationStackCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var listOperationsOperationSpec = {
    httpMethod: "GET",
    path: "providers/Microsoft.Web/operations",
    queryParameters: [
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.CsmOperationCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var getAvailableStacksOnPremOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/providers/Microsoft.Web/availableStacks",
    urlParameters: [
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.osTypeSelected,
        Parameters.apiVersion0
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationStackCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var getAvailableStacksNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationStackCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var listOperationsNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.CsmOperationCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
var getAvailableStacksOnPremNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationStackCollection
        },
        default: {
            bodyMapper: Mappers.DefaultErrorResponse
        }
    },
    serializer: serializer
};
