// @azure/arm-apimanagement/esm/operations/apiExport.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/apiExportMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ApiExport. */
var ApiExport = /** @class */ (function () {
    /**
     * Create a ApiExport.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function ApiExport(client) {
        this.client = client;
    }
    ApiExport.prototype.get = function (resourceGroupName, serviceName, apiId, format, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            apiId: apiId,
            format: format,
            options: options
        }, getOperationSpec, callback);
    };
    return ApiExport;
}());
export { ApiExport };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var getOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.apiId0,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.format1,
        Parameters.exportParameter,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ApiExportResult
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
