// @azure/arm-apimanagement/esm/operations/apiManagementServiceSkus.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/apiManagementServiceSkusMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ApiManagementServiceSkus. */
var ApiManagementServiceSkus = /** @class */ (function () {
    /**
     * Create a ApiManagementServiceSkus.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function ApiManagementServiceSkus(client) {
        this.client = client;
    }
    ApiManagementServiceSkus.prototype.listAvailableServiceSkus = function (resourceGroupName, serviceName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            options: options
        }, listAvailableServiceSkusOperationSpec, callback);
    };
    ApiManagementServiceSkus.prototype.listAvailableServiceSkusNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listAvailableServiceSkusNextOperationSpec, callback);
    };
    return ApiManagementServiceSkus;
}());
export { ApiManagementServiceSkus };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listAvailableServiceSkusOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/skus",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ResourceSkuResults
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
var listAvailableServiceSkusNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ResourceSkuResults
        },
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
