// @azure/arm-apimanagement/esm/operations/apiProduct.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/apiProductMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ApiProduct. */
var ApiProduct = /** @class */ (function () {
    /**
     * Create a ApiProduct.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function ApiProduct(client) {
        this.client = client;
    }
    ApiProduct.prototype.listByApis = function (resourceGroupName, serviceName, apiId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            apiId: apiId,
            options: options
        }, listByApisOperationSpec, callback);
    };
    ApiProduct.prototype.listByApisNext = function (nextPageLink, options, callback) {
        return this.client.sendOperationRequest({
            nextPageLink: nextPageLink,
            options: options
        }, listByApisNextOperationSpec, callback);
    };
    return ApiProduct;
}());
export { ApiProduct };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByApisOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/products",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.apiId1,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.filter0,
        Parameters.top,
        Parameters.skip,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ProductCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var listByApisNextOperationSpec = {
    httpMethod: "GET",
    baseUrl: "https://management.azure.com",
    path: "{nextLink}",
    urlParameters: [
        Parameters.nextPageLink
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ProductCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
