// @azure/arm-apimanagement/esm/operations/notificationRecipientEmail.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/notificationRecipientEmailMappers";
import * as Parameters from "../models/parameters";
/** Class representing a NotificationRecipientEmail. */
var NotificationRecipientEmail = /** @class */ (function () {
    /**
     * Create a NotificationRecipientEmail.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function NotificationRecipientEmail(client) {
        this.client = client;
    }
    NotificationRecipientEmail.prototype.listByNotification = function (resourceGroupName, serviceName, notificationName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            notificationName: notificationName,
            options: options
        }, listByNotificationOperationSpec, callback);
    };
    NotificationRecipientEmail.prototype.checkEntityExists = function (resourceGroupName, serviceName, notificationName, email, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            notificationName: notificationName,
            email: email,
            options: options
        }, checkEntityExistsOperationSpec, callback);
    };
    NotificationRecipientEmail.prototype.createOrUpdate = function (resourceGroupName, serviceName, notificationName, email, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            notificationName: notificationName,
            email: email,
            options: options
        }, createOrUpdateOperationSpec, callback);
    };
    NotificationRecipientEmail.prototype.deleteMethod = function (resourceGroupName, serviceName, notificationName, email, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            notificationName: notificationName,
            email: email,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    return NotificationRecipientEmail;
}());
export { NotificationRecipientEmail };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByNotificationOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/notifications/{notificationName}/recipientEmails",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.notificationName,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RecipientEmailCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var checkEntityExistsOperationSpec = {
    httpMethod: "HEAD",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/notifications/{notificationName}/recipientEmails/{email}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.notificationName,
        Parameters.email,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        204: {},
        404: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/notifications/{notificationName}/recipientEmails/{email}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.notificationName,
        Parameters.email,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.RecipientEmailContract
        },
        201: {
            bodyMapper: Mappers.RecipientEmailContract
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/notifications/{notificationName}/recipientEmails/{email}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.notificationName,
        Parameters.email,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
