// @azure/arm-apimanagement/esm/operations/quotaByCounterKeys.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as tslib_1 from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/quotaByCounterKeysMappers";
import * as Parameters from "../models/parameters";
/** Class representing a QuotaByCounterKeys. */
var QuotaByCounterKeys = /** @class */ (function () {
    /**
     * Create a QuotaByCounterKeys.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function QuotaByCounterKeys(client) {
        this.client = client;
    }
    QuotaByCounterKeys.prototype.listByService = function (resourceGroupName, serviceName, quotaCounterKey, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            quotaCounterKey: quotaCounterKey,
            options: options
        }, listByServiceOperationSpec, callback);
    };
    QuotaByCounterKeys.prototype.update = function (resourceGroupName, serviceName, quotaCounterKey, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            quotaCounterKey: quotaCounterKey,
            parameters: parameters,
            options: options
        }, updateOperationSpec, callback);
    };
    return QuotaByCounterKeys;
}());
export { QuotaByCounterKeys };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var listByServiceOperationSpec = {
    httpMethod: "GET",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/quotas/{quotaCounterKey}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.quotaCounterKey,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {
            bodyMapper: Mappers.QuotaCounterCollection
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
var updateOperationSpec = {
    httpMethod: "PATCH",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/quotas/{quotaCounterKey}",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.quotaCounterKey,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: tslib_1.__assign({}, Mappers.QuotaCounterValueContractProperties, { required: true })
    },
    responses: {
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
