// @azure/arm-apimanagement/esm/operations/userConfirmationPassword.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/userConfirmationPasswordMappers";
import * as Parameters from "../models/parameters";
/** Class representing a UserConfirmationPassword. */
var UserConfirmationPassword = /** @class */ (function () {
    /**
     * Create a UserConfirmationPassword.
     * @param {ApiManagementClientContext} client Reference to the service client.
     */
    function UserConfirmationPassword(client) {
        this.client = client;
    }
    UserConfirmationPassword.prototype.send = function (resourceGroupName, serviceName, userId, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            serviceName: serviceName,
            userId: userId,
            options: options
        }, sendOperationSpec, callback);
    };
    return UserConfirmationPassword;
}());
export { UserConfirmationPassword };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var sendOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/users/{userId}/confirmations/password/send",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.serviceName,
        Parameters.userId,
        Parameters.subscriptionId
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        204: {},
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
