// @azure/arm-appservice/esm/models/parameters.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
export var acceptLanguage = {
    parameterPath: "acceptLanguage",
    mapper: {
        serializedName: "accept-language",
        defaultValue: 'en-US',
        type: {
            name: "String"
        }
    }
};
export var analysisName = {
    parameterPath: "analysisName",
    mapper: {
        required: true,
        serializedName: "analysisName",
        type: {
            name: "String"
        }
    }
};
export var apiVersion0 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2018-02-01',
        type: {
            name: "String"
        }
    }
};
export var apiVersion1 = {
    parameterPath: "apiVersion",
    mapper: {
        required: true,
        isConstant: true,
        serializedName: "api-version",
        defaultValue: '2018-11-01',
        type: {
            name: "String"
        }
    }
};
export var backupId = {
    parameterPath: "backupId",
    mapper: {
        required: true,
        serializedName: "backupId",
        type: {
            name: "String"
        }
    }
};
export var baseAddress = {
    parameterPath: "baseAddress",
    mapper: {
        required: true,
        serializedName: "baseAddress",
        type: {
            name: "String"
        }
    }
};
export var billingLocation = {
    parameterPath: [
        "options",
        "billingLocation"
    ],
    mapper: {
        serializedName: "billingLocation",
        type: {
            name: "String"
        }
    }
};
export var certificateOrderName = {
    parameterPath: "certificateOrderName",
    mapper: {
        required: true,
        serializedName: "certificateOrderName",
        type: {
            name: "String"
        }
    }
};
export var deletedSiteId = {
    parameterPath: "deletedSiteId",
    mapper: {
        required: true,
        serializedName: "deletedSiteId",
        type: {
            name: "String"
        }
    }
};
export var deleteEmptyServerFarm = {
    parameterPath: [
        "options",
        "deleteEmptyServerFarm"
    ],
    mapper: {
        serializedName: "deleteEmptyServerFarm",
        type: {
            name: "Boolean"
        }
    }
};
export var deleteMetrics = {
    parameterPath: [
        "options",
        "deleteMetrics"
    ],
    mapper: {
        serializedName: "deleteMetrics",
        type: {
            name: "Boolean"
        }
    }
};
export var detailed = {
    parameterPath: [
        "options",
        "detailed"
    ],
    mapper: {
        serializedName: "detailed",
        type: {
            name: "Boolean"
        }
    }
};
export var details = {
    parameterPath: [
        "options",
        "details"
    ],
    mapper: {
        serializedName: "details",
        type: {
            name: "Boolean"
        }
    }
};
export var detectorName = {
    parameterPath: "detectorName",
    mapper: {
        required: true,
        serializedName: "detectorName",
        type: {
            name: "String"
        }
    }
};
export var diagnosticCategory = {
    parameterPath: "diagnosticCategory",
    mapper: {
        required: true,
        serializedName: "diagnosticCategory",
        type: {
            name: "String"
        }
    }
};
export var diagnosticsName = {
    parameterPath: "diagnosticsName",
    mapper: {
        required: true,
        serializedName: "diagnosticsName",
        type: {
            name: "String"
        }
    }
};
export var domainName0 = {
    parameterPath: "domainName",
    mapper: {
        required: true,
        serializedName: "domainName",
        type: {
            name: "String"
        }
    }
};
export var domainName1 = {
    parameterPath: "domainName",
    mapper: {
        required: true,
        serializedName: "domainName",
        constraints: {
            Pattern: /[a-zA-Z0-9][a-zA-Z0-9\.-]+/
        },
        type: {
            name: "String"
        }
    }
};
export var domainOwnershipIdentifierName = {
    parameterPath: "domainOwnershipIdentifierName",
    mapper: {
        required: true,
        serializedName: "domainOwnershipIdentifierName",
        type: {
            name: "String"
        }
    }
};
export var durationInSeconds = {
    parameterPath: [
        "options",
        "durationInSeconds"
    ],
    mapper: {
        serializedName: "durationInSeconds",
        type: {
            name: "Number"
        }
    }
};
export var endTime0 = {
    parameterPath: [
        "options",
        "endTime"
    ],
    mapper: {
        serializedName: "endTime",
        type: {
            name: "DateTime"
        }
    }
};
export var endTime1 = {
    parameterPath: [
        "options",
        "endTime"
    ],
    mapper: {
        serializedName: "endTime",
        type: {
            name: "String"
        }
    }
};
export var entityName = {
    parameterPath: "entityName",
    mapper: {
        required: true,
        serializedName: "entityName",
        type: {
            name: "String"
        }
    }
};
export var environmentName = {
    parameterPath: "environmentName",
    mapper: {
        required: true,
        serializedName: "environmentName",
        type: {
            name: "String"
        }
    }
};
export var expiredOnly = {
    parameterPath: [
        "options",
        "expiredOnly"
    ],
    mapper: {
        serializedName: "expiredOnly",
        type: {
            name: "Boolean"
        }
    }
};
export var featured = {
    parameterPath: [
        "options",
        "featured"
    ],
    mapper: {
        serializedName: "featured",
        type: {
            name: "Boolean"
        }
    }
};
export var filter = {
    parameterPath: [
        "options",
        "filter"
    ],
    mapper: {
        serializedName: "$filter",
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export var forceDelete = {
    parameterPath: [
        "options",
        "forceDelete"
    ],
    mapper: {
        serializedName: "forceDelete",
        type: {
            name: "Boolean"
        }
    }
};
export var forceHardDeleteDomain = {
    parameterPath: [
        "options",
        "forceHardDeleteDomain"
    ],
    mapper: {
        serializedName: "forceHardDeleteDomain",
        type: {
            name: "Boolean"
        }
    }
};
export var functionName = {
    parameterPath: "functionName",
    mapper: {
        required: true,
        serializedName: "functionName",
        type: {
            name: "String"
        }
    }
};
export var gatewayName = {
    parameterPath: "gatewayName",
    mapper: {
        required: true,
        serializedName: "gatewayName",
        type: {
            name: "String"
        }
    }
};
export var hostingEnvironmentName = {
    parameterPath: "hostingEnvironmentName",
    mapper: {
        required: true,
        serializedName: "hostingEnvironmentName",
        type: {
            name: "String"
        }
    }
};
export var hostName0 = {
    parameterPath: [
        "options",
        "hostName"
    ],
    mapper: {
        serializedName: "hostName",
        type: {
            name: "String"
        }
    }
};
export var hostName1 = {
    parameterPath: "hostName",
    mapper: {
        required: true,
        serializedName: "hostName",
        type: {
            name: "String"
        }
    }
};
export var id = {
    parameterPath: "id",
    mapper: {
        required: true,
        serializedName: "id",
        type: {
            name: "String"
        }
    }
};
export var includeSlots = {
    parameterPath: [
        "options",
        "includeSlots"
    ],
    mapper: {
        serializedName: "includeSlots",
        type: {
            name: "Boolean"
        }
    }
};
export var instance = {
    parameterPath: "instance",
    mapper: {
        required: true,
        serializedName: "instance",
        type: {
            name: "String"
        }
    }
};
export var instanceId = {
    parameterPath: "instanceId",
    mapper: {
        required: true,
        serializedName: "instanceId",
        type: {
            name: "String"
        }
    }
};
export var linuxDynamicWorkersEnabled = {
    parameterPath: [
        "options",
        "linuxDynamicWorkersEnabled"
    ],
    mapper: {
        serializedName: "linuxDynamicWorkersEnabled",
        type: {
            name: "Boolean"
        }
    }
};
export var linuxWorkersEnabled = {
    parameterPath: [
        "options",
        "linuxWorkersEnabled"
    ],
    mapper: {
        serializedName: "linuxWorkersEnabled",
        type: {
            name: "Boolean"
        }
    }
};
export var location = {
    parameterPath: "location",
    mapper: {
        required: true,
        serializedName: "location",
        type: {
            name: "String"
        }
    }
};
export var maxFrameLength = {
    parameterPath: [
        "options",
        "maxFrameLength"
    ],
    mapper: {
        serializedName: "maxFrameLength",
        type: {
            name: "Number"
        }
    }
};
export var name = {
    parameterPath: "name",
    mapper: {
        required: true,
        serializedName: "name",
        type: {
            name: "String"
        }
    }
};
export var namespaceName = {
    parameterPath: "namespaceName",
    mapper: {
        required: true,
        serializedName: "namespaceName",
        type: {
            name: "String"
        }
    }
};
export var nextPageLink = {
    parameterPath: "nextPageLink",
    mapper: {
        required: true,
        serializedName: "nextLink",
        type: {
            name: "String"
        }
    },
    skipEncoding: true
};
export var operationId = {
    parameterPath: "operationId",
    mapper: {
        required: true,
        serializedName: "operationId",
        type: {
            name: "String"
        }
    }
};
export var osType = {
    parameterPath: [
        "options",
        "osType"
    ],
    mapper: {
        serializedName: "osType",
        type: {
            name: "String"
        }
    }
};
export var osTypeSelected = {
    parameterPath: [
        "options",
        "osTypeSelected"
    ],
    mapper: {
        serializedName: "osTypeSelected",
        type: {
            name: "String"
        }
    }
};
export var premierAddOnName = {
    parameterPath: "premierAddOnName",
    mapper: {
        required: true,
        serializedName: "premierAddOnName",
        type: {
            name: "String"
        }
    }
};
export var processId = {
    parameterPath: "processId",
    mapper: {
        required: true,
        serializedName: "processId",
        type: {
            name: "String"
        }
    }
};
export var propertiesToInclude = {
    parameterPath: [
        "options",
        "propertiesToInclude"
    ],
    mapper: {
        serializedName: "propertiesToInclude",
        type: {
            name: "String"
        }
    }
};
export var publicCertificateName = {
    parameterPath: "publicCertificateName",
    mapper: {
        required: true,
        serializedName: "publicCertificateName",
        type: {
            name: "String"
        }
    }
};
export var recommendationId = {
    parameterPath: [
        "options",
        "recommendationId"
    ],
    mapper: {
        serializedName: "recommendationId",
        type: {
            name: "String"
        }
    }
};
export var relayName = {
    parameterPath: "relayName",
    mapper: {
        required: true,
        serializedName: "relayName",
        type: {
            name: "String"
        }
    }
};
export var resourceGroupName = {
    parameterPath: "resourceGroupName",
    mapper: {
        required: true,
        serializedName: "resourceGroupName",
        constraints: {
            MaxLength: 90,
            MinLength: 1,
            Pattern: /^[-\w\._\(\)]+[^\.]$/
        },
        type: {
            name: "String"
        }
    }
};
export var routeName = {
    parameterPath: "routeName",
    mapper: {
        required: true,
        serializedName: "routeName",
        type: {
            name: "String"
        }
    }
};
export var sasUrl = {
    parameterPath: [
        "options",
        "sasUrl"
    ],
    mapper: {
        serializedName: "sasUrl",
        type: {
            name: "String"
        }
    }
};
export var siteExtensionId = {
    parameterPath: "siteExtensionId",
    mapper: {
        required: true,
        serializedName: "siteExtensionId",
        type: {
            name: "String"
        }
    }
};
export var siteName = {
    parameterPath: "siteName",
    mapper: {
        required: true,
        serializedName: "siteName",
        type: {
            name: "String"
        }
    }
};
export var skipToken = {
    parameterPath: [
        "options",
        "skipToken"
    ],
    mapper: {
        serializedName: "$skipToken",
        type: {
            name: "String"
        }
    }
};
export var sku = {
    parameterPath: [
        "options",
        "sku"
    ],
    mapper: {
        serializedName: "sku",
        type: {
            name: "String"
        }
    }
};
export var slot = {
    parameterPath: "slot",
    mapper: {
        required: true,
        serializedName: "slot",
        type: {
            name: "String"
        }
    }
};
export var snapshotId = {
    parameterPath: "snapshotId",
    mapper: {
        required: true,
        serializedName: "snapshotId",
        type: {
            name: "String"
        }
    }
};
export var softRestart = {
    parameterPath: [
        "options",
        "softRestart"
    ],
    mapper: {
        serializedName: "softRestart",
        type: {
            name: "Boolean"
        }
    }
};
export var sourceControlType = {
    parameterPath: "sourceControlType",
    mapper: {
        required: true,
        serializedName: "sourceControlType",
        type: {
            name: "String"
        }
    }
};
export var startTime0 = {
    parameterPath: [
        "options",
        "startTime"
    ],
    mapper: {
        serializedName: "startTime",
        type: {
            name: "DateTime"
        }
    }
};
export var startTime1 = {
    parameterPath: [
        "options",
        "startTime"
    ],
    mapper: {
        serializedName: "startTime",
        type: {
            name: "String"
        }
    }
};
export var subscriptionId = {
    parameterPath: "subscriptionId",
    mapper: {
        required: true,
        serializedName: "subscriptionId",
        type: {
            name: "String"
        }
    }
};
export var subscriptionName = {
    parameterPath: "subscriptionName",
    mapper: {
        required: true,
        serializedName: "subscriptionName",
        type: {
            name: "String"
        }
    }
};
export var synchronous = {
    parameterPath: [
        "options",
        "synchronous"
    ],
    mapper: {
        serializedName: "synchronous",
        type: {
            name: "Boolean"
        }
    }
};
export var threadId = {
    parameterPath: "threadId",
    mapper: {
        required: true,
        serializedName: "threadId",
        type: {
            name: "String"
        }
    }
};
export var timeGrain0 = {
    parameterPath: [
        "options",
        "timeGrain"
    ],
    mapper: {
        serializedName: "timeGrain",
        constraints: {
            Pattern: /PT[1-9][0-9]+[SMH]/
        },
        type: {
            name: "String"
        }
    }
};
export var timeGrain1 = {
    parameterPath: [
        "options",
        "timeGrain"
    ],
    mapper: {
        serializedName: "timeGrain",
        type: {
            name: "String"
        }
    }
};
export var top = {
    parameterPath: [
        "options",
        "top"
    ],
    mapper: {
        serializedName: "$top",
        type: {
            name: "String"
        }
    }
};
export var updateSeen = {
    parameterPath: [
        "options",
        "updateSeen"
    ],
    mapper: {
        serializedName: "updateSeen",
        type: {
            name: "Boolean"
        }
    }
};
export var view = {
    parameterPath: "view",
    mapper: {
        required: true,
        serializedName: "view",
        type: {
            name: "String"
        }
    }
};
export var vnetName = {
    parameterPath: "vnetName",
    mapper: {
        required: true,
        serializedName: "vnetName",
        type: {
            name: "String"
        }
    }
};
export var webJobName = {
    parameterPath: "webJobName",
    mapper: {
        required: true,
        serializedName: "webJobName",
        type: {
            name: "String"
        }
    }
};
export var workerName = {
    parameterPath: "workerName",
    mapper: {
        required: true,
        serializedName: "workerName",
        type: {
            name: "String"
        }
    }
};
export var workerPoolName = {
    parameterPath: "workerPoolName",
    mapper: {
        required: true,
        serializedName: "workerPoolName",
        type: {
            name: "String"
        }
    }
};
export var xenonWorkersEnabled = {
    parameterPath: [
        "options",
        "xenonWorkersEnabled"
    ],
    mapper: {
        serializedName: "xenonWorkersEnabled",
        type: {
            name: "Boolean"
        }
    }
};
